### PR TITLE
feat: sync job to update reval placements

### PR DESCRIPTION
**Pull request by CaiWillis**
TIS21-4830

### CODE REVIEW

**Comment by CaiWillis** [comment]
RevalTisDateChangeSync drawio

**Comment by CaiWillis** [comment]
Motivation:
The code for getting the Placements would be identical to what we already have for programme memberships except for the base query - I've consolidated it all into a more generic abstract template, otherwise the sonar code duplication stat would be through the roof

**Review by hihilary — COMMENTED** [review]
I think placement and programme membership are a little different here.
For programme membership, we export the current pm for every doctor.
But for placement, we have to remove/add doctors when their F1 grade starts/ends.

In another word, when a doctor's F1 grade placement starts, we get nothing from the TCS SQL, how do you know which record to be removed from elasticsearch?

**Comment by CaiWillis** [comment]
> 

Thanks for this, will go back and have a think

**Comment by CaiWillis** [comment]
Update: This PR now has the context of us sending the placement grade over to reval, no longer any need to add/remove doctors, only update their placement grades

**Comment by hihilary** [comment]
We've got the current programme membership date change job on the sync service UI, as the image shown below. Are you considering adding the placement one? 
image

File: src/main/resources/config/application.yml [hunk 1]
```diff
@@ -223,6 +223,7 @@ application:
     reval:
       exchange: ${REVAL_RABBIT_EXCHANGE:reval.exchange}
       routingKey.currentpm.update: ${REVAL_RABBIT_CURRENTPMUPDATE_ROUTING_KEY:reval.currentpm.update}
+      routingKey.currentplacement.update: ${REVAL_RABBIT_CURRENTPLACEMENTUPDATE_ROUTING_KEY:reval.currentplacement.update}
```
Comment (hihilary) [inline src/main/resources/config/application.yml:226]: I couldn't start up the service locally. Got an error: `Could not resolve placeholder 'application.rabbit.reval.routingKey.trainee.update' in value "${application.rabbit.reval.routingKey.trainee.update}"`

**Review by hihilary — CHANGES_REQUESTED** [review]

File: src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    this.rabbitMqPublisher = rabbitMqPublisher;
+  }
+
+  @Override
+  public void run(String params) {
+    revalCurrentPlacementSyncJob();
+  }
+
+  @Profile("!nimdta")
+  @Scheduled(cron = "${application.cron.revalCurrentPlacementJob}")
```
Comment (hihilary) [inline src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java:44]: Could not resolve placeholder 'application.cron.revalCurrentPlacementJob' in value "${application.cron.revalCurrentPlacementJob}"

Comment (CaiWillis) [reply src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java:44]: Thanks for spotting these, they're covered in the TIS-OPS PR but I forgot to do them here 🙈

File: src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import org.springframework.jmx.export.annotation.ManagedOperation;
+import org.springframework.jmx.export.annotation.ManagedResource;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+import uk.nhs.tis.sync.job.PersonDateChangeCaptureSyncJobTemplate;
+import uk.nhs.tis.sync.message.publisher.RabbitMqTcsRevalTraineeUpdatePublisher;
+
+/**
+ * Get personIds whose current programmeMembership changes nightly. And sends messages to rabbitMq
+ * for tcs to fetch
```
Comment (hihilary) [inline src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java:21]: programmeMembership -> placement

File: src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ */
+@Profile("!nimdta")
+@Component
+@ManagedResource(objectName = "sync.mbean:name=RevalCurrentPlacementSyncJob",
+    description = "Job message personIds if their programme membership(s) started/ended")
```
Comment (hihilary) [inline src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java:26]: the same as above.

File: src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      String endDate = dateOfChange.minusDays(1).format(DateTimeFormatter.ISO_LOCAL_DATE);
+      return BASE_QUERY.replace(":endDate", endDate).replace(":startDate", startDate)
+          .replace(":pageSize", "" + DEFAULT_PAGE_SIZE);
+    } else {
+      return BASE_QUERY.replace(":pageSize", "" + DEFAULT_PAGE_SIZE)
+          .replace(" AND (dateTo = ':endDate' OR dateFrom = ':startDate')", "");
+    }
+  }
+
+
```
Comment (hihilary) [inline src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java:81]: Blank lines should be removed.

File: src/main/resources/static/index.html [hunk 6]
```diff
@@ -46,7 +46,7 @@ <h2 class="box-1">Person Record Status Sync</h2>
   </div>
 </div>
 <br>
-<div id="reval-panel" class="hidden">
+<div id="reval-panel" class="">
```
Comment (hihilary) [inline src/main/resources/static/index.html:49]: The panel is default to be `hidden`, and only when profile doesn't contain nimdta, `hidden` is removed.
Please check the method `fetchSysProfile()` in `main.js`.

**Review by hihilary — CHANGES_REQUESTED** [review]

File: src/main/resources/static/index.html (hunk 6 above)
Comment (CaiWillis) [reply src/main/resources/static/index.html:49]: Whoops, that's just from local testing

**Review by hihilary — CHANGES_REQUESTED** [review]
I haven't found any fix for this comment: https://github.com/Health-Education-England/TIS-SYNC/pull/238#discussion_r1319604596

**Comment by CaiWillis** [comment]
> I haven't found any fix for this comment: [#238 (comment)](https://github.com/Health-Education-England/TIS-SYNC/pull/238#discussion_r1319604596)

I have removed this unused parameter

File: src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+ */
+@Profile("!nimdta")
+@Component
+@ManagedResource(objectName = "sync.mbean:name=RevalCurrentPlacementSyncJob",
+    description = "Job message personIds if their placement(s) started/ended")
+public class RevalCurrentPlacementSyncJob extends PersonDateChangeCaptureSyncJobTemplate<Long> {
+  private static final String BASE_QUERY =
+      "SELECT DISTINCT traineeId FROM Placement" + " WHERE traineeId > :lastPersonId"
+          + " AND (dateFrom = ':endDate' OR DateTo = ':startDate')"
+          + " ORDER BY personId LIMIT :pageSize";
```
Comment (hihilary) [inline src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPlacementSyncJob.java:31]: When running this job locally, I got an exception `SQLGrammarException: could not extract ResultSet`. I think it's caused by `personId` not found.

File: src/main/resources/config/application.yml [hunk 8]
```diff
@@ -222,7 +223,7 @@ application:
   rabbit:
     reval:
       exchange: ${REVAL_RABBIT_EXCHANGE:reval.exchange}
-      routingKey.currentpm.update: ${REVAL_RABBIT_CURRENTPMUPDATE_ROUTING_KEY:reval.currentpm.update}
+      routingKey.trainee.update: ${REVAL_RABBIT_TRAINEEUPDATE_ROUTING_KEY:reval.trianee.update}
```
Comment (hihilary) [inline src/main/resources/config/application.yml:226]: There's a typo in the routing key name.

**Review by hihilary — CHANGES_REQUESTED** [review]

**Review by hihilary — DISMISSED** [review]

**Review by hihilary — APPROVED** [review]

**Pull request merged**